cpp/src/join.cpp

```diff
@@ -200,8 +204,185 @@ bool is_repartition_not_needed(const TaskContext& ctx,
+    right_fields.emplace_back(key);
+  }
+  arrow::acero::HashJoinNodeOptions join_opts(legate_to_arrow_join_type(join_type),
+                                              left_fields,
+                                              right_fields,
+                                              arrow::compute::literal(true),
+                                              "_left",
+                                              "_right");
+
+  arrow::acero::Declaration left{"table_source", arrow::acero::TableSourceNodeOptions{lhs}};
+  arrow::acero::Declaration right{"table_source", arrow::acero::TableSourceNodeOptions{rhs}};
+  arrow::acero::Declaration hashjoin{"hashjoin", {left, right}, std::move(join_opts)};
+
+  return ARROW_RESULT(arrow::acero::DeclarationToTable(std::move(hashjoin)));
+}
+
+// Creating some constant for each arrow type is surprisingly hard
+std::shared_ptr<arrow::Scalar> create_default_value_scalar(
+  const std::shared_ptr<arrow::DataType>& type)
+{
+  // Constructing a string/binary type with 0 is a runtime error
+  // Catch this and try constructing with a string
+  try {
+    return ARROW_RESULT(arrow::MakeScalar(type, 0));
+  } catch (const std::exception& e) {
+    return ARROW_RESULT(arrow::MakeScalar(type, std::string{}));
+  }
+}
+
+// Here we have to do some hacking because arrow does not support nulls in joins
+// The general approach is to take the original key and demote it to a value
+// Then take a copy of this demoted column, replace the nulls with a sentinel value
+// Perform the join and then remove the extra columns
+std::shared_ptr<arrow::Table> arrow_join_and_gather_with_null(
+  TaskContext& ctx,
+  std::shared_ptr<arrow::Table> lhs,
+  std::shared_ptr<arrow::Table> rhs,
+  const std::vector<std::string> lhs_keys,
+  const std::vector<std::string> rhs_keys,
+  JoinType join_type)
+{
+  auto lhs_temp = lhs->Slice(0);
+  auto rhs_temp = rhs->Slice(0);
+  // Create a copy of each key column and replace nulls with a sentinel value
+  std::vector<arrow::FieldRef> left_fields;
+  for (const auto& key : lhs_keys) {
+    auto key_array = lhs->GetColumnByName(key);
+    auto scalar    = create_default_value_scalar(key_array->type());
+    auto key_replaced_nulls =
+      ARROW_RESULT(arrow::compute::CallFunction("coalesce", {key_array, scalar})).chunked_array();
+    lhs_temp =
+      ARROW_RESULT(lhs_temp->AddColumn(lhs_temp->num_columns(),
+                                       arrow::field(key + "_replaced", key_replaced_nulls->type()),
+                                       key_replaced_nulls));
+    left_fields.emplace_back(arrow::FieldRef(key + "_replaced"));
+  }
+
+  std::vector<arrow::FieldRef> right_fields;
+  for (const auto& key : rhs_keys) {
+    auto key_array = rhs->GetColumnByName(key);
+    auto scalar    = create_default_value_scalar(key_array->type());
+    auto key_replaced_nulls =
+      ARROW_RESULT(arrow::compute::CallFunction("coalesce", {key_array, scalar})).chunked_array();
```

Comment:
I guess it's the same, but let's use:
```suggestion
      ARROW_RESULT(arrow::compute::CallFunction("fill_null", {key_array, scalar})).chunked_array();
```